yarpl/include/yarpl/flowable/Subscriber.h

```diff
@@ -15,40 +15,96 @@ namespace flowable {
 template <typename T>
 class Subscriber : public virtual Refcounted, public yarpl::enable_get_ref {
  public:
+  virtual void onSubscribe(Reference<Subscription>) = 0;
+  virtual void onComplete() = 0;
+  virtual void onError(folly::exception_wrapper) = 0;
+  virtual void onNext(T) = 0;
+};
+
+// codemod all things that inherit from Subscriber to inherit from
+// LegacySubscriber
+template <typename T>
+class LegacySubscriber : public Subscriber<T> {
+ public:
   // Note: If any of the following methods is overridden in a subclass, the new
   // methods SHOULD ensure that these are invoked as well.
-  virtual void onSubscribe(Reference<Subscription> subscription) {
+  void onSubscribe(Reference<Subscription> subscription) override {
     DCHECK(subscription);
+    CHECK(!subscription_);
+    subscription_ = std::move(subscription);
+  }
+
+  // No further calls to the subscription after this method is invoked.
+  void onComplete() override {
+    DCHECK(subscription_) << "Calling onComplete() without a subscription";
+    subscription_.reset();
+  }
 
-    if (subscription_) {
-      subscription->cancel();
-      return;
-    }
+  // No further calls to the subscription after this method is invoked.
+  void onError(folly::exception_wrapper) override {
+    DCHECK(subscription_) << "Calling onError() without a subscription";
+    subscription_.reset();
+  }
 
+ protected:
+  Reference<Subscription> subscription() {
+    return subscription_;
+  }
+
+ private:
+  Reference<Subscription> subscription_;
+};
+
+template <typename T>
+class SafeSubscriber : public Subscriber<T> {
+ public:
+  // Note: If any of the following methods is overridden in a subclass, the new
+  // methods SHOULD ensure that these are invoked as well.
+  void onSubscribe(Reference<Subscription> subscription) final override {
+    DCHECK(subscription);
+    CHECK(!subscription_);
     subscription_ = std::move(subscription);
+    onSubscribeImpl();
   }
 
   // No further calls to the subscription after this method is invoked.
-  virtual void onComplete() {
+  void onComplete() final override {
     DCHECK(subscription_) << "Calling onComplete() without a subscription";
     subscription_.reset();
```

Comment:
I think this needs to be `auto sub = std::move(subscription_);` or `auto sub = std::atomic_exchange(&subscription_, nullptr);` in a thread-safe world.

The last reference to the `Subscriber` could be the `Subscription`, so we don't want to kill it just yet.  The easy test case is to make a cycle between a `Subscriber` and a `Subscription`, grab a raw pointer to the `Subscriber`, leak it, and then call `rawSubscriber->onComplete();`.

Reply:
Spec states that signaling methods won't be called without external synchronization, so we shouldn't have any races here. 

However you're right, we might kill ourselves when we reset the subscription object, so let's hold it on the stack until we're done.

Reply:
Actually it's slightly worse than that; a terminating signal method might do something that destroys `this` (a bug we've seen often); if this is SafeSubscriber we should keep a reference to ourselves on the stack.